packages/contracts-bedrock/src/dispute/PrestateRegistry.sol

```diff
@@ -0,0 +1,152 @@
+    Timestamp internal immutable L2_GENESIS_BLOCK_TIMESTAMP;
+    /// @notice The L2 genesis block.
+    uint256 internal immutable L2_GENESIS_BLOCK;
+    /// @notice The L2 chain id.
+    uint256 internal immutable L2_CHAIN_ID;
+    /// @notice The L2 block time.
+    uint256 internal immutable L2_BLOCK_TIME;
+
+    /// @notice A list of L2 hardforks, in ascending order. This list does not have to be exhaustive, but it must be
+    ///         sorted and contain upgrades that alter the Fault Proof Program prestate hashes.
+    Hardfork[] public hardforks;
+
+    constructor(
+        SuperchainConfig _superchainConfig,
+        Timestamp _l2GenesisBlockTimestamp,
+        uint256 _l2GenesisBlock,
+        uint256 _l2ChainId,
+        uint256 _l2BlockTime
+    ) {
+        L2_CHAIN_ID = _l2ChainId;
+        L2_GENESIS_BLOCK = _l2GenesisBlock;
+        L2_GENESIS_BLOCK_TIMESTAMP = _l2GenesisBlockTimestamp;
+        L2_BLOCK_TIME = _l2BlockTime;
+        SUPERCHAIN_CONFIG = _superchainConfig;
+    }
+
+    /// @notice Registers a hardfork with the prestate registry.
+    /// @param _l2ActivationTime The L2 timestamp at which the hardfork is activated.
+    /// @param _prestates The prestates for the hardfork.
+    function registerHardfork(Timestamp _l2ActivationTime, PrestateInformation[] memory _prestates) external {
+        // INVARIANT: Only the guardian can register hardforks.
+        if (msg.sender != SUPERCHAIN_CONFIG.guardian()) revert BadAuth();
+
+        // Convert the L2 timestamp to an L2 block number.
+        uint256 activationBlock = l2TimestampToBlock(_l2ActivationTime);
+
+        uint256 currentLength = hardforks.length;
+        if (currentLength > 0) {
+            Hardfork storage latestFork = hardforks[currentLength - 1];
+
+            // INVARIANT: Hardforks must be registered in ascending order.
+            if (latestFork.l2ActivationBlock >= activationBlock) revert OutOfOrderHardfork();
+        }
+
+        // Register all of the prestates for the hardfork.
+        hardforks.push();
+        for (uint256 i = 0; i < _prestates.length; i++) {
+            PrestateInformation memory prestate = _prestates[i];
+            hardforks[currentLength].prestates[prestate.vmID][prestate.programID] = prestate.prestateHash;
+        }
+        hardforks[currentLength].l2ActivationBlock = activationBlock;
+    }
+
+    /// @notice Revokes the latest hardfork, if it has not happened yet.
+    function revokePendingFork() external {
+        // INVARIANT: Only the guardian can revoke pending hardforks.
+        if (msg.sender != SUPERCHAIN_CONFIG.guardian()) revert BadAuth();
+
+        // INVARIANT: If the fork has already been activated, it cannot be revoked.
+        // NOTE: This check is imperfect as we reference based on L1 time, but this drift is acceptable.
+        Hardfork storage latestFork = hardforks[hardforks.length - 1];
+        if (l2BlockToTimestamp(latestFork.l2ActivationBlock).raw() <= block.timestamp) {
```

Comment:
Consider adding a buffer here, to account for the drift. Depends on the timestamp drift that the L2 allows though.

Reply:
I think it's tight enough that we probably don't need a buffer. This doesn't need to be perfectly precise, and we'd likely not revoke a fork minutes before it happened.